python/llm/example/GPU/Speculative-Decoding/Eagle/evaluation/gen_ea_answer_llama2chat.py

```diff
@@ -0,0 +1,500 @@
+        model.past_key_values_data = past_key_values_data
+        model.current_length_data = current_length_data
+
+    input_len = input_ids.shape[1]
+    reset_tree_mode(model)
+    tree_logits, logits, hidden_state, sample_token = initialize_tree(
+        input_ids, model, tree_buffers["tree_attn_mask"], past_key_values, logits_processor
+    )
+    new_token = 0
+
+    for idx in range(max_steps):
+        candidates, cart_candidates_prob, tree_candidates = generate_candidates(
+            tree_logits,
+            tree_buffers["tree_indices"],
+            tree_buffers["retrieve_indices"],
+            sample_token,
+            logits_processor
+        )
+        logits, hidden_state_new, outputs = tree_decoding(
+            model,
+            tree_candidates,
+            past_key_values,
+            tree_buffers["tree_position_ids"],
+            input_ids,
+            tree_buffers["retrieve_indices_head"],
+        )
+        best_candidate, accept_length, sample_p = evaluate_posterior(
+            logits, candidates, logits_processor, cart_candidates_prob, tree_logits[2], tree_buffers["p_indices"],
+            tree_candidates, tree_buffers["b_indices"]
+        )
+        input_ids, tree_logits, new_token, hidden_state, sample_token = update_inference_inputs(
+            input_ids,
+            candidates,
+            best_candidate,
+            accept_length,
+            tree_buffers["retrieve_indices"],
+            logits_processor,
+            logits,
+            tree_logits,
+            new_token,
+            past_key_values_data,
+            current_length_data,
+            model,
+            hidden_state,
+            hidden_state_new,
+            sample_p
+        )
+        if tokenizer.eos_token_id in input_ids[0, input_len:].tolist():
+            break
+        if new_token > 1024:
+            break
+        if input_ids.shape[1] > 1960:
+            break
+    return input_ids, new_token, idx
+
+
+def run_eval(
+        base_model_path,
+        ea_model_path,
+        model_id,
+        question_file,
+        question_begin,
+        question_end,
+        answer_file,
+        max_new_token,
+        num_choices,
+        num_gpus_per_model,
+        num_gpus_total,
+        max_gpu_memory,
+        temperature,
+        tree_choices,
+        enable_ipex_llm,
+):
+    questions = load_questions(question_file, question_begin, question_end)
+    shuffled_ids = [q["question_id"] for q in questions]
+
+    # Split the question file into `num_gpus` files
+    assert num_gpus_total % num_gpus_per_model == 0
+    use_ray = num_gpus_total // num_gpus_per_model > 1
+
+    if use_ray:
+        get_answers_func = ray.remote(num_gpus=num_gpus_per_model)(
+            get_model_answers
+        ).remote
+    else:
+        get_answers_func = get_model_answers
+
+    chunk_size = len(questions) // (num_gpus_total // num_gpus_per_model)  # // 2
+    ans_handles = []
```

Comment:
remove multiple gpu related stuff if we didn't test on multiple xpus

Reply:
done